Initial implementation of virtfs-ext3

movitto opened the pull request:
Includes sequential modifications to existing codebase to:

- bring it up to standards
- implement the virtfs interface
- add specs

movitto commented:
@roliveri this should be ready for review

roliveri commented:
@hsong-rh @jerryk55 Please join in on the review.

jerryk55 commented on lib/virtfs/ext3/directory.rb:
```diff
@@ -1,51 +1,71 @@
+      @sb        = sb
+      @inode_num = inode_num
+      @inode_obj = sb.get_inode(inode_num)
+      @data      = sb.ext4? ? @inode_obj.read : FileData.new(@inode_obj, @sb).read
```
I'm curious as to why we would continue handling this is an Ext3 filesystem if it is actually an Ext4 filesystem.  Shouldn't this be an error and/or an exception?

jerryk55 commented on lib/virtfs/ext3/directory.rb:
```diff
@@ -1,51 +1,71 @@
-      @data = FileData.new(@inodeObj, @sb).read
+    def initialize(sb, inode_num = ROOT_DIRECTORY)
+      raise "nil superblock"   if sb.nil?
+      raise "nil inode number" if inode_num.nil?
```
Is the assumption here that the raised exception will display the Module::Class.method info?  Why else would you remove this helpful info?

jerryk55 commented on lib/virtfs/ext3/alloc_bitmap.rb:
```diff
@@ -1,16 +1,16 @@
   class AllocBitmap
     def initialize(data)
-      raise "Ext3::AllocBitmap.initialize: Nil data" if data.nil?
+      raise "Ext::AllocBitmap.initialize: Nil data" if data.nil?
```
Why did you remove the "3"?

movitto replied:
I was thinking so that there errors would be more generic / applicable to both the Ext3 and Ext4 modules. Not a big deal, can change it back if that is preferred

jerryk55 commented on lib/virtfs/ext3/directory.rb:
```diff
@@ -1,51 +1,85 @@
+      @inode_num = inode_num
+      @inode_obj = sb.get_inode(inode_num)
+      @data      = sb.ext4? ? @inode_obj.read : FileData.new(@inode_obj, @sb).read
+    end
```
If the superblock is Ext4, isn't that an error or an exception?  I wouldn't think we would just continue to run the initialization as if nothing happened.

movitto replied:
@jerryk55 the Ext4::Inode class provides a 'read' method which we can use to access the inode data directly, but not such method exists in the ext3 module. This was added so that both modules can leverage the common Directory code.

At some point the "read" method should be added to the Ext3::inode class for consistency.

jerryk55 commented on lib/virtfs/ext3/directory.rb:
```diff
@@ -1,51 +1,85 @@
         break if @data[p, 4].nil?
-        de = DirectoryEntry.new(@data[p..-1], newEnt)
-        raise "Ext3::Directory.globEntries: DirectoryEntry length cannot be 0" if de.len == 0
+        de = DirectoryEntry.new(fs, @data[p..-1], new_entry)
```
Why did you add a new "fs" argument?  It is not used in the DirectoryEntry class except to save it, and I can find no other code that makes use of the accessor method.

movitto replied:
The 'fs' argument was added to comply with the 'virtfs' interface which is expecting it in that accessor on Directory instances. Without it, virtfs will throw an error when it tries to call the 'fs' method

jerryk55 commented on lib/virtfs/ext3/directory_entry.rb:
```diff
@@ -31,37 +31,65 @@ class DirectoryEntry
-      out += "Len     : #{len}\n"
-      out += "Name len: 0x#{'%04x' % @de['name_len']}\n"
-      out += "Type    : #{fileType}\n" if @isNew
-      out += "Name    : #{name}\n"
```
The dump methods are very helpful when attempting to debug SSA issues. I would lobby to put them back (I haven't noticed if you removed any other ones).

movitto replied:
As mentioned earlier, the dump methods were removed at the same time as the references to $log. I'll readd the dump methods with the next PR revision, though we need add a central logging mechanism to the virtfs module

movitto replied:
@jerryk55 dump methods readded & standardized

jerryk55 commented on lib/virtfs/ext3/file_data.rb:
```diff
@@ -68,17 +68,17 @@ def write(buf, _len = buf.length)
-      startBlock, startByte = pos.divmod(@blockSize)
-      endBlock, endByte = (pos + len - 1).divmod(@blockSize)
+      startBlock, startByte = pos.divmod(@block_size)
+      endBlock, endByte = (pos + len - 1).divmod(@block_size)
```
If you're going through the incredibly tedious chore of aligning things and getting rid of camel-case when possible maybe the "=" here should line up with the previous line?

jerryk55 commented on lib/virtfs/ext3/group_descriptor_entry.rb:
```diff
@@ -31,35 +31,20 @@ def initialize(buf)
+    def num_dirs
       @gde['num_dirs']
     end
-
```
See previous comment about dump methods.

jerryk55 commented on lib/virtfs/ext3/group_descriptor_table.rb:
```diff
@@ -32,23 +32,11 @@ def [](group)
       @gdt[group]
     end
 
-    def dump(dump_bitmaps = false)
```
Ditto.

jerryk55 commented:
@movitto it looks like the tests are failing because something can't find virtfs.

jerryk55 commented on lib/virtfs/ext3/superblock.rb:
```diff
@@ -208,199 +208,111 @@ def gdt
       field & bit == bit
     end
-
-    # Dump object.
```
Yet another ditto to the previous comments on dump.  You get the picture.  :-)

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+      f.atime
+    end
+
+    def file_blockdev?(p)
```
Not yet implemented?

movitto replied:
Yes, a few features in the new virtfs interface were never implemented in MiQFS, since these pull requests are for extraction / refactoring / testing, I did not make any functional changes / add new features

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+      f.mtime
+    end
+
+    def file_owned?(p)
```
Several more unimplemented methods here?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+    def file_grpowned?(p)
+    end
+
+    def file_identical?(p1, p2)
```
Not yet implemented?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+    def file_ftype(p)
+    end
+
+    def file_grpowned?(p)
```
Not yet implemented?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+      !f.nil? && f.file?
+    end
+
+    def file_ftype(p)
```
Not yet implemented?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+    def file_executable?(p)
+    end
+
+    def file_executable_real?(p)
```
Not yet implemented?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+      !f.nil? && f.dir?
+    end
+
+    def file_executable?(p)
```
Not yet implemented?

jerryk55 commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+    def file_blockdev?(p)
+    end
+
+    def file_chardev?(p)
```
Not yet implemented?

movitto commented:
@jerryk55 see comments above (updated PR will be sent tomorrow)

movitto commented:
@jerryk55 @roliveri pushed the updated PR, a few changes

- the patchset has been simplied to 1. refactoring 2. virtfs interface 3. specs

- the 'dump' methods were readded and other outstanding comments addressed

- more style changes were added for consistency

movitto commented:
As a side note, we discovered a potential bug w/ the current ext3 module today that needs further exploration (and potential fixing both here and in the main miq codebase):

At line 93 of the current "file_data.rb" module, we seemingly retrieve the triple indirect block using the double indirect pointer. This needs investigating & a possible fix.

https://github.com/ManageIQ/virtfs-ext3/blob/master/lib/virtfs/ext3/file_data.rb#L99

jerryk55 commented on spec/factories/ext.rb:
```diff
@@ -0,0 +1,13 @@
+
+FactoryGirl.define do
+  factory :ext, class: OpenStruct do
+    path '/home/mmorsi/workspace/cfme/virtfs-ext3/ext3.fs'
```
Just noticed this path.  Obviously you want something else here...

jerryk55 commented on lib/virtfs/ext3/base.rb:
```diff
@@ -0,0 +1,50 @@
+
+    def initialize(blk_device)
+      blk_device.seek(0, IO::SEEK_SET)
+      @superblock  = Superblock.new(blk_device)
```
Just noticed that you invoke "Superblock.new" here but on line 12 above the call is "VirtFS::Ext3::Superblock.new".
If this line works then the more explicit line is probably not necessary, right?

movitto replied:
@jerryk55 simplied w/ the new update

jerryk55 commented on lib/virtfs/ext3.rb:
```diff
@@ -1 +1,16 @@
 require "virtfs/ext3/version"
+require_relative 'ext3/alloc_bitmap'
```
See @Fryguy comment on my PR https://github.com/ManageIQ/virtfs-xfs/pull/1 - "Minor, but it's common in gems to just refer to file assuming lib is on the path. That is, instead of require_relative 'xfs/whatever', prefer require 'virtfs/xfs/whatever'".  If this is desired we should be consistent throughout all the VirtFS and VirtFS-fstype gems.

movitto replied:
@jerryk55 @Fryguy reading through the xfs PR, it seems that we are going w/ 'require_relative' correct? And thus no additional work is needed for this

roliveri replied:
@movitto Correct.

movitto commented:
@roliveri @jerryk55 this should be pretty close to go. That being said the camcorderfs integration still isn't working 100% as intended. The recorder is wired up and the cassettes written but I don't believe the right transactions are being recorded & playback is failing ('disk not found' once it is moved).

Does anything look off w/ the initialization code? (in spec_helper)

jerryk55 commented on lib/virtfs/ext3/superblock.rb:
```diff
@@ -137,270 +139,309 @@ class Superblock
+      @num_groups, @last_group_blocks = @sb['num_blocks'].divmod(@sb['blocks_in_group'])
+
+      @num_groups += 1 if @last_group_blocks > 0
+      @fsId        = DiskUUID.parse_raw(@sb['fs_id'])
```
I assume that DiskUUID isn't really defined anywhere?

jerryk55 commented on lib/virtfs/ext3/superblock.rb:
```diff
@@ -1,19 +1,16 @@
 
 require 'binary_struct'
-require 'util/miq-uuid'
+require 'virt_disk/disk_uuid'
```
This file doesn't exist, right?  I noticed that you added a dependency to the UUIDTools gem in the Gemfile - is your intent to change this to use that other gem?

jerryk55 commented on Gemfile:
```diff
@@ -1,5 +1,8 @@
 source 'https://rubygems.org'
 
+gem 'virtfs', "~> 0.0.1", :git => "https://github.com/ManageIQ/virtfs.git", :branch => "master"
+gem 'comcorderfs', "~> 0.1.0", :git => "https://github.com/ManageIQ/virtfs-camcorderfs.git", :branch => "master"
```
Typo in 'comcorderfs', right?

movitto commented:
@jerryk55 ah yes, I had those components in my virt_disk checkout locally but never committed them. Updated virt_disk PR #1 to include them.

Also fixed the typo, good catch!

roliveri commented:
@movitto I think I might know the cause of your test issues with CamcorderFS.
The tests don't activate VirtFS, explicitly referencing `VirtFS::VFile, etc`, but I'm not sure what `VirtDisk::BlockFile` is doing. If it's calling `File.open` instead of `VirtFS::Vfile.open` then the IO from the disk file won't be recorded.

If that's the case, there are 2 possible solutions:
1 - Activate `VirtFS` in the tests.
2 - Change the version of `BlockFile` that the tests use to use `VirtFS::VFile`.

Option #2 would probably be the easiest solution. I'd just create a test version of `BlockFile` in the spec directory and use that.

movitto commented:
@roliveri I just updated the virt_disk PR to check if VirtFS is defined and if so use VirtFS::VFile, else use the existing File.open mechanism. Unfortunately while some progress was made (some tests broke due to the referencing of an incorrect mount point which I fixed), the cassettes still do not contain the relevant recordings.

Pushed the updated virtfs-ext3 and virt_disk pr in the hopes of getting more eyeballs on this to sort it out.

movitto commented:
@roliveri also I checked and I believe the virtfs-ext3 module is compliant w/ how we are approaching the project layout for these components. If I'm wrong or missed anything just shout out and I can fix

movitto commented:
@roliveri @jerryk55  success! The addition to virt_disk/block_file was needed as per your suggestion rich, but I had to do the opposite of your other suggestion and leave set the camcorder recorder root when mounting that device (perhaps this should be done internally in camcorderfs). 

https://github.com/ManageIQ/virtfs-ext3/pull/1/commits/7346f2c9fd3595138ed551508609560c30161dc1#diff-23dc0ec957f4bb1eba766196149f8004R19

I noticed that when it wasn't there more disk interactions were being recorded (as a result of the addition to virt_disk) but they were incorrect. After this change camcorder was able to pickup all the operations on the ext3 image being done through virtfs.

Small fix but took a while to discover. Good thing is though I verified that the specs fully run from the cassettes when the disk image is moved, proving the camcorderfs concept, and providing an example for further implementations.

Should be just about ready to merge!

roliveri commented:
@movitto Ah, I forgot I suggested removing setting root - yeah, that was just an initial shot in the dark, and was wrong. Sorry my suggestion caused you to waste so much time.

roliveri commented:
Now the question is, why can't the test find camcorders?

movitto commented:
@roliveri @jerryk55 the camcorder based tests are now working and this PR should be ready to go. Besides various issues w/ deps I had to add a fixed path to the factory used in the specs so that travis could run against consistent targets. We may want to change this at some point, but other than that this is done

roliveri commented on lib/virtfs/ext3/base.rb:
```diff
@@ -0,0 +1,50 @@
+module VirtFS::Ext3
+  class FS
```
Following the file naming convention, this file should be: `virtues/ext3/fs.rb`.

roliveri commented on lib/virtfs/ext3/dir_class.rb:
```diff
@@ -0,0 +1,74 @@
+require 'active_support/core_ext/object/try' # until we can use the safe nav operator
+
+module VirtFS::Ext3
+  class FS
```
Because re-opening the class to extend it breaks the file naming conventions, we've updated code like this to implement the methods in a module, which can then be included in the class in question.

So here, you'd implement the `DirClassMethods` module under the `FS` class:
```ruby
module VirtFS::Ext3
  class FS
    module DirCalssMethods
    :
    end
  end
end
```

and the file would then be: `virtfs/ext3/fs/dir_class_methods.rb`.
The class would then be included in the FS class.

movitto replied:
@roliveri updated

roliveri commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+module VirtFS::Ext3
+  class FS
```
Same comment here as for the Dir class methods:

Because re-opening the class to extend it breaks the file naming conventions, we've updated code like this to implement the methods in a module, which can then be included in the class in question.

So here, you'd implement the `FileClassMethods` module under the `FS` class:
```ruby
module VirtFS::Ext3
  class FS
    module FileCalssMethods
    :
    end
  end
end
```

and the file would then be: `virtfs/ext3/fs/file_class_methods.rb`.
The class would then be included in the FS class.

movitto replied:
@roliveri updated

roliveri commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+      def get_file(p)
+        p = unnormalize_path(p)
+
+        dir, fname = ::File.split(p)
```
Remember under VirtFS the ::File and ::Dir classes can be reassigned. If you need to ensure you're using the core class, you should use `VfsRealFile`.

movitto replied:
@roliveri updated

roliveri commented on lib/virtfs/ext3/file_class.rb:
```diff
@@ -0,0 +1,188 @@
+    def file_writable_real?(p)
+    end
+
+    def file_new(f, parsed_args, _open_path, _cwd)
```
The `file_new` method needs to return a `File` class instance that implements the required file instance methods for the thin interface (see `virtfs/nativefs/thin/file.rb`). This appears to return a directory entry object, that doesn't appear to implement the required methods - so I'm not sure how this can work.

movitto replied:
@roliveri ah this is because DirectoryEntry implements all the required methods, in the existing ext3 code which this was based off of, File just dispatches to that

https://github.com/ManageIQ/manageiq-gems-pending/blob/master/lib/gems/pending/fs/MiqFS/modules/Ext3.rb#L32

Regardless, I updated the patch to wrap the dir entry in a File instance and added methods to the File class delegating calls to the encapsulated DE

roliveri commented on lib/virtfs/ext3/superblock.rb:
```diff
@@ -1,19 +1,16 @@
-  # ////////////////////////////////////////////////////////////////////////////
-  # // Data definitions. Linux 2.6.2 from Fedora Core 6.
-
+module VirtFS::Ext3
```
Should the following constants be defined under the `SuperBlock` class instead of dumping them in the `Ext3` class?

movitto replied:
@roliveri  moved these under superblock

roliveri commented on spec/dir_class_spec.rb:
```diff
@@ -0,0 +1,188 @@
+      expect(VirtFS::VDir["*"]).to match_array([])
+    end
+
+    it "should enumerate the same file names as the standard Dir.glob - simple glob" do
```
I can see how tests like this work for the NativeFS modules, because the same files can be accessed through the standard interface as well as VirtFS, but how does this work for the Ext3 module? The files on the ext filesystem aren't accessible through the standard interface.

movitto replied:
@roliveri not fully following, these tests are testing access to the mounted filesystem (the actual mounting occurs in 'before' hooks in spec_helper.rb).

roliveri replied:
@movitto Maybe the confusion is in the test doesn't seem to be testing what it states:
`"should enumerate the same file names as the standard Dir.glob - simple glob"`
It's not calling the standard `Dir.glob`, so how can it be testing that they return the same file names?

In the VirtFS tests, the mounted filesystem is NativeFS, so the same files and directories are visible through VirtFS as well as through the standard Ruby interface, so we can compare the return values. For Ext3, the files and directories ore only visible through VirtFS, so you can't compare the return values with the standard call.

movitto commented:
@roliveri  updated, see feedback above

movitto commented:
@roliveri let me know if there is anything else

movitto commented:
@roliveri I just started a new PR w/ everything split out in a manner that makes sense to the current state of the changes. I'm hoping it'll be easy to review & I got all your feedback.

The spec issue is still outstanding though, I believe this is just a misnomer as it isn't verifying the virtfs results against a runtime 'Dir.glob' but rather the glob output that I generated locally. We could probably rename these test cases to something like 'it should enumerate the same file names as the native ruby mechanisms' or similar.
